Uniform tracing

daumayr opened the pull request:
This PR adds uniform record and replay for all concurrency models featured in SOMns.

### Threads & Locks
Record and Replay for Locking works as usual, see CSP.
For signal and await operations the lock has to be held before the operation can be performed -> already synchronised.

The tricky part is awaitTimeout and lock reacquisitions after await returns.
awaitTimeout is handled by recording whether a given await timed out, if it timed out, we don't perform it in replay. if awaitTimeout returned before the timeout, we use a regular await to avoid nasty timing issues.

For the lock reacquisitions the easiest solution to implement (also avoids duplicating and messing with a lot of low-level implementation stuff of reentrant lock) is to use a special internal replay condition for each lock. when await returns, we check if the locks version number is right, if not we await on the internal condition. This means the lock is released again so that other threads can perform operations that have to happen before the awaiting thread continues.
The internal condition is signalled when the version number is changed, delayed wakeup from await then checks again, and may perform another await if the version is still not right.

### CEL Actors
Reimplemented the record and replay for actors with the new approach.

### STM
STM replay is based on recording the order of successful commits. global commit lock is reused for synchronization, version number only incremented on successful commits.

daumayr committed:
implement TL replay

daumayr committed:
Implement Actor + STM uniform rr

daumayr committed:
port memory tracing per iteration

daumayr committed:
add Poisoning of actors in replay

Co-Authored-By: Stefan Marr <git@stefan-marr.de>

daumayr commented on tests/replay/test.sh:
```diff
@@ -30,15 +36,35 @@ else
+    "Vacation 1 0 7 4"
   )
+
+  declare -a STM=(
```
make this singlethreaded, we know there is an issue

daumayr commented on src/som/interpreter/actors/Actor.java:
```diff
@@ -340,7 +350,8 @@ protected void execute(final ForkJoinPool actorPool) {
   @Override
-  public void setStepToNextTurn(final boolean val) {}
+  public void setStepToNextTurn(final boolean val) {
+  }
```
eclipseformat issue

daumayr commented on src/som/interpreter/transactions/Transactions.java:
```diff
@@ -39,8 +44,10 @@
 
-  private Transactions() {}
+  private Transactions() {
+  }
```
eclipse format error

daumayr commented on src/som/primitives/processes/ChannelPrimitives.java:
```diff
@@ -112,7 +111,8 @@ public ActivityType getType() {
 
-    protected void beforeExec(final SInvokable disp) {}
+    protected void beforeExec(final SInvokable disp) {
+    }
```
eclipse format issue

daumayr commented on src/som/primitives/ActivitySpawn.java:
```diff
@@ -202,7 +218,8 @@ protected boolean hasTagIgnoringEagerness(final Class<? extends Tag> tag) {
     @Child protected ExceptionSignalingNode notAValue;
 
-    @Child RecordOneEvent traceProcCreation = new RecordOneEvent(TraceRecord.PROCESS_CREATION);
+    @Child RecordOneEvent traceProcCreation =
```
make conditional

daumayr commented on src/som/primitives/ActivitySpawn.java:
```diff
@@ -117,7 +133,7 @@ public static IsValue createIsValue() {
     @Child protected ExceptionSignalingNode notAValue;
     @Child RecordOneEvent                   traceProcCreation =
-        new RecordOneEvent(TraceRecord.PROCESS_CREATION);
+        new RecordOneEvent(TraceRecord.ACTIVITY_CREATION);
```
make conditional

daumayr commented on src/som/primitives/threading/ConditionPrimitives.java:
```diff
@@ -39,12 +45,24 @@ public final Condition doCondition(final Condition cond) {
   @Primitive(primitive = "threadingAwait:")
   public abstract static class AwaitPrim extends UnaryExpressionNode {
+
+    @Child protected static RecordOneEvent traceWakeup =
```
make conditional

daumayr commented on src/som/primitives/threading/ConditionPrimitives.java:
```diff
@@ -39,12 +45,24 @@ public final Condition doCondition(final Condition cond) {
+          tc.owner.replayIncrementEventNo();
+        }
+
+        // Output.println("Activity " + aid + " woke up");
```
remove comment

daumayr commented on src/som/primitives/threading/ConditionPrimitives.java:
```diff
@@ -57,14 +75,35 @@ public final Condition doCondition(final Condition cond) {
+    @Child protected static RecordOneEvent traceTimeout =
+        new RecordOneEvent(TraceRecord.CONDITION_TIMEOUT);
+    @Child protected static RecordOneEvent traceWakeup  =
+        new RecordOneEvent(TraceRecord.CONDITION_WAKEUP);
```
make conditional

daumayr commented on src/som/primitives/threading/MutexPrimitives.java:
```diff
@@ -14,23 +14,53 @@
   @Primitive(primitive = "threadingLock:", selector = "lock")
   public abstract static class LockPrim extends UnaryExpressionNode {
+    @Child protected static RecordOneEvent traceLock =
+        new RecordOneEvent(TraceRecord.LOCK_LOCK);
```
mkae conditional

daumayr commented on src/som/primitives/threading/MutexPrimitives.java:
```diff
@@ -14,23 +14,53 @@
+        if (VmSettings.ACTOR_TRACING) {
+          ((TracingLock) lock).tracingLock(traceLock);
+          // Output.println(
+          // "" + aid + " acquired lock v" + (((TracingLock) lock).getNextEventNumber() - 1));
```
remove comments

daumayr commented on src/som/primitives/threading/MutexPrimitives.java:
```diff
@@ -86,16 +116,29 @@ public Object critical(final ReentrantLock lock, final SBlock block) {
   @Primitive(primitive = "threadingIsLocked:")
   public abstract static class IsLockedPrim extends UnaryExpressionNode {
+    @Child protected RecordOneEvent traceIsLocked =
+        new RecordOneEvent(TraceRecord.LOCK_ISLOCKED);
```
make conditional

daumayr commented on src/som/vm/Activity.java:
```diff
@@ -67,7 +83,8 @@ default int getNextTraceBufferId() {
    * Does nothing for non-tracing activities, i.e., when debugging is disabled.
    */
-  default void setStepToJoin(final boolean val) {}
+  default void setStepToJoin(final boolean val) {
```
eclipse format issue

daumayr commented on src/som/vm/ObjectSystem.java:
```diff
@@ -502,7 +501,8 @@ private int handlePromiseResult(final SPromise promise) {
       try {
         Thread.sleep(500);
-      } catch (InterruptedException e) {}
+      } catch (InterruptedException e) {
```
eclipse format issue

daumayr commented on src/tools/concurrency/TracingActors.java:
```diff
@@ -170,7 +186,6 @@ public ReplayActor(final VM vm) {
-        expectedMessages = vm.getTraceParser().getExpectedMessages(activityId);
         replayEvents = vm.getTraceParser().getReplayEventsForEntity(activityId);
 
         synchronized (actorList) {
```
check if this is still used

daumayr commented on src/tools/concurrency/TracingActors.java:
```diff
@@ -408,30 +258,22 @@ private MessageRecord peekExpected() {
         if (mailboxExtension != null) {
           for (EventualMessage msg : mailboxExtension) {
-            if (!progress && a.replayCanProcess(a.nextExpectedMessage, msg)) {
+            if (!progress && a.replayCanProcess(msg)) {
```
double check the progress part of the condition

daumayr replied:
postponedmessages should be a priority queue?

daumayr commented on src/tools/replay/TraceParser.java:
```diff
@@ -231,7 +227,8 @@ protected void parseTrace(final boolean scanning, final Subtrace loc,
       }
-    } catch (IOException e) {
+    } catch (Throwable e) {
+      e.printStackTrace();
```
undo this change

daumayr commented on src/tools/replay/TraceParser.java:
```diff
@@ -315,69 +303,33 @@ private boolean readTrace(final boolean first, final boolean scanning, final Byt
-        assert b.position() == start + RecordEventNodes.TWO_EVENT_SIZE;
-        break;
       default:
+        Output.println("MISSING");
```
remove

daumayr commented on src/tools/replay/nodes/TraceContextNode.java:
```diff
@@ -15,26 +15,24 @@
   protected void recordGeneric(final Activity activity) {
-    writeContext(activity.getId(), activity.getNextTraceBufferId(),
-        TraceRecord.ACTOR_CONTEXT.value);
+    writeContext(activity.getId(), activity.getNextTraceBufferId());
```
add specializations for the other activities

daumayr committed:
fixes and cleanup

smarr commented on .gitlab-ci.yml:
```diff
@@ -63,6 +63,15 @@ benchmark_savina_csp_job:
+  allow_failure: true
+  script:
+    - ant compile
+    - export EXP=`if [[ "$CI_BUILD_REF_NAME" = "master" ]]; then echo "SOMns-Mutex"; else echo "SOMns-Mutex-exp"; fi`; rebench --without-nice -c --commit-id="$CI_BUILD_REF" --environment="`hostname` `cat /etc/issue | cut -d ' ' -f 1`" --project=SOMns-Mutex --branch=master codespeed.conf $EXP
```
```suggestion
    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-Mutex
```

smarr commented on core-lib/Benchmarks/AsyncHarness.ns:
```diff
@@ -120,9 +120,18 @@ class AsyncHarness usingPlatform: platform = Value (
+      | stats = system traceStatistics.
+        memstats = system memoryStatistics.
+      |
       (name + ': trace size:    ' + (stats at: 1) + 'byte') println.
```
Can we also have these statistics conditionally?

smarr commented on core-lib/Benchmarks/Harness.ns:
```diff
@@ -145,6 +145,18 @@ class Harness usingPlatform: platform = Value (
+      | stats = system traceStatistics.
+        memstats = system memoryStatistics.
+      |
+      (name + ': trace size:    ' + (stats at: 1) + 'byte') println.
```
Can we also have these statistics conditionally?

smarr commented on core-lib/System.ns:
```diff
@@ -48,6 +48,8 @@ class System usingVmMirror: vmMirror = Value (
 
   (* Actor Tracing Statistics *)
   public traceStatistics = ( ^ vmMirror traceStatistics: nil )
```
Can we have a `#traceStatistics`?

daumayr replied:
do you mean hasTraceStatistics prim and method so that its only printed when useful?

smarr replied:
yes a hasTraceStatistics prim would be great for consistency

smarr commented on src/som/Launcher.java:
```diff
@@ -58,14 +57,6 @@ private static void finalizeExecution(final int exitCode) {
-    }
-
-    if (VmSettings.MEMORY_TRACING) {
-      TracingBackend.reportPeakMemoryUsage();
```
This can be derived from the `memoryStatistics`, right?

daumayr replied:
yes, there is now a primitive to get the data, and printing is handled in the harnesses

smarr commented on src/som/interpreter/actors/EventualSendNode.java:
```diff
@@ -325,6 +339,14 @@ public final Object toNearRefWithoutResultPromise(final Object[] args) {
       }
 
+      if (VmSettings.REPLAY) {
+        // TODO similar thing for any other usages of actor.send(). especially in timer prim
```
What's this comment about? perhaps needs to be rephrased? What is "similar thing" referring to?

daumayr replied:
before send is used, the send event needs to be consumed and the version number attached to the message. So if we have e.g. an http server that sends messages as callbacks, it has to do that. Exceptions to this rule are when a message already has the version number stored in it, e.g. due to some promise scenarios.

smarr replied:
hmm, I get your explanation, but not how it relates to the comment. What exactly needs to be implemented still?

daumayr replied:
Essentially it's a note on what to do when implementing external event stuff. maybe I should remove the comment and make a note somewhere else

smarr replied:
Perhaps it can be rephrased to something like:
`// Note: When implementing external events, such as ... ...`

smarr commented on src/som/interpreter/actors/RegisterOnPromiseNode.java:
```diff
@@ -25,14 +37,57 @@ public void register(final SPromise promise, final PromiseMessage msg,
+            npr = current.getNextReplayEvent();
+
+            assert msg instanceof AbstractPromiseSendMessage;
+            LinkedList<ReplayRecord> events = new LinkedList<>();
```
Any specific reason why this is a `LinkedList`?

daumayr replied:
Looked it up, anything that maintains order is fine.

smarr reviewed (commented):
Initial comments

sophie-kaleba commented on codespeed.conf:
```diff
@@ -713,6 +739,16 @@ experiments:
+            - SOMns-graal-tn:
+                suites:
+                  - mutex
+
```
Aren't those same benchmarks already ran a few lines above (see SOMns-Mutex)?

daumayr replied:
the exp stuff was for the CI benchmarking in brussels, with results reported to the speed website. I guess that was deprecated by stefan and I should change it.

smarr replied:
Yeah, I removed the `exp` duplication, with the data going to ReBenchDB, that's not needed any longer.

sophie-kaleba commented on src/som/interpreter/actors/Actor.java:
```diff
@@ -252,8 +262,8 @@ void doRun() {
       t.currentlyExecutingActor = actor;
 
-      if (VmSettings.ACTOR_TRACING) {
-        ActorExecutionTrace.recordActivityContext((TracingActor) actor, tracer);
```
Was this cast to `TracingActor` unnecessary in the first place? Otherwise I didn't catch a change that would explain why it is not needed anymore

daumayr replied:
The tracer nodes were changed, they are now specialising on the activity type and are no longer restricted to actors. of course this is not visible here. I probably should refactor ActorExecution trace, the name isn't accurate anymore

sophie-kaleba commented on src/som/vm/VmSettings.java:
```diff
@@ -63,7 +63,7 @@
     TRACE_SMALL_IDS = getBool("som.smallIds", false);
 
-    ACTOR_TRACING = getBool("som.actorTracing", false);
+    UNIFORM_TRACING = getBool("som.actorTracing", false);
```
Shouldn't the setting's name be turned into `som.uniformTracing` as well?

daumayr replied:
yes, changing the flags etc would be a good idea. so from -at to -ut

daumayr committed:
implement uniform receiver side actor rr

daumayr committed:
fix after rebase

daumayr committed:
add bench conf

daumayr committed:
fix missing fj activity context

daumayr committed:
add multi paradigm to setup

daumayr committed:
fix using wrong vm setting

daumayr committed:
print trace stats only when needed

daumayr committed:
Refactor ActorExecutionTrace to reflect uniformness

daumayr committed:
rephrase comment

daumayr committed:
remove outdated exp-runs

daumayr committed:
update test-data

daumayr committed:
fix style

daumayr committed:
fix eclipse format

daumayr committed:
archive traces on failed test

smarr commented on src/tools/replay/TraceRecord.java:
```diff
@@ -4,22 +4,25 @@
   MESSAGE(2),
   PROMISE_MESSAGE(3),
   SYSTEM_CALL(4),
-  CHANNEL_CREATION(5),
```
What's happening now with `CHANNEL_CREATION`?
That's not needed anymore, because it is deterministic?
Was this an artifact from the Kompos tracing for which we want it?

smarr replied:
Is not needed for replay anymore, because they don't need their own ids.

smarr commented on tests/replay/test.sh:
```diff
@@ -30,15 +44,35 @@ else
-    "DeadLock 100 0 4:2:3"
-    "Messages 100 0 1000"
-    "Sequence 100 0 100"
+    #Currently broken due to timerthread being incompatible
```
Can I help with this, where would I start?

smarr reviewed (commented):
some notes for the moment. Working through the rest still.

smarr committed:
Fix rebase

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix codespeed.conf

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@daumayr what's the difference between the bench.conf and the codespeed.conf? do we need both?

For the codespeed.conf, I fixed some issue already, but mutex experiments look like they are not working at the moment, and might have some copy and paste mistake or so:

```
Running Benchmarks: 0.00%
 Running Benchmarks:               Philosophers (SOMns-interp-tn, mutex, 500 4) 1 None None	mean:       55.2	time left: 00:00:13: 25.00%
 Running Benchmarks:                Philosophers (SOMns-graal-tn, mutex, 500 4) 1 None None	mean:      292.0	time left: 00:00:13: 50.00%
 Executing run:ProducerConsumer (SOMns-interp-tn, mutex, 4000 4) 1 None None
     cmd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns/som -G   core-lib/Benchmarks/Harness.ns SavinaCSP.ProducerConsumer 50 0 4000 4
     cwd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns
     Run failed. Return code: 1
     Output:
         MessageNotUnderstood(SavinaCSP>>#ProducerConsumer)
         
     Execution has failed, benchmark is aborted.
         The benchmark failed 1 times in a row.
 Running Benchmarks:          ProducerConsumer (SOMns-interp-tn, mutex, 4000 4) 1 None None	mean:        0.0	time left: 00:00:04: 75.00%
 Executing run:ProducerConsumer (SOMns-graal-tn, mutex, 4000 4) 1 None None
     cmd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns/som  core-lib/Benchmarks/Harness.ns SavinaCSP.ProducerConsumer 50 0 4000 4
     cwd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns
     Run failed. Return code: 1
     Output:
         MessageNotUnderstood(SavinaCSP>>#ProducerConsumer)
         
     Execution has failed, benchmark is aborted.
         The benchmark failed 1 times in a row.
 Running Benchmarks:           ProducerConsumer (SOMns-graal-tn, mutex, 4000 4) 1 None None	mean:        0.0	time left: 00:00:00: 100.00%
 --------------------------------------------------------------------------------------------------
   Benchmark          Executor          Suite   Extra    Core   Size   Var   #Samples   Mean (ms)  
 --------------------------------------------------------------------------------------------------
   ProducerConsumer   SOMns-graal-tn    mutex   4000 4      1                       0   Failed     
   Philosophers       SOMns-graal-tn    mutex   500 4       1                      20   292        
   ProducerConsumer   SOMns-interp-tn   mutex   4000 4      1                       0   Failed     
   Philosophers       SOMns-interp-tn   mutex   500 4       1                      20   55         
 --------------------------------------------------------------------------------------------------
```

smarr commented on .gitlab-ci.yml:
```diff
@@ -63,6 +63,15 @@ benchmark_savina_csp_job:
+    - ant compile
+    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-Mutex
+
+
```
```suggestion
    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-SavinaCSP
    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-Mutex

```

Let's avoid recompiling things. Both jobs should be short enough.

smarr commented on codespeed.conf:
```diff
@@ -60,13 +60,13 @@ benchmark_suites:
                 codespeed_name: "1st.Vacation"
             - VacationSTM:
-                extra_args: "1 0 8"
+                extra_args: "1 0 1"
```
Are these changes needed?
I would prefer to keep parameters constant, if possible.
Otherwise, we can't compare over time.

daumayr replied:
The STM version is slower, the idea was to speed things up

smarr commented on codespeed.conf:
```diff
@@ -594,6 +594,22 @@ benchmark_suites:
+    mutex:
+        description: various mutex based benchmarks
+        gauge_adapter: RebenchLog
+        command: *CSP_INTERP_CMD
```
Is this the correct command?

smarr commented on src/tools/replay/nodes/RecordEventNodes.java:
```diff
@@ -1,13 +1,12 @@
   public static final int ONE_EVENT_SIZE   = 1 + Long.BYTES;
-  public static final int TWO_EVENT_SIZE   = 1 + (2 * Long.BYTES);
-  public static final int THREE_EVENT_SIZE = 1 + (3 * Long.BYTES);
+  public static final int THREE_EVENT_SIZE = 1 + Long.BYTES + Long.BYTES + Long.BYTES;
```
This is used in only one place in the `TraceParser`. Looks like something that is not used anymore?

smarr commented on src/tools/replay/nodes/TraceContextNode.java:
```diff
@@ -2,11 +2,13 @@
+import tools.replay.actors.UniformExecutionTrace.UniformTraceBuffer;
 
 
 public abstract class TraceContextNode extends TraceNode {
```
This looks like something that could be simplified with a `ClassProfile` instead of the various specializations.
It seems we really just need the concrete type to resolve the virtual method calls, right?

smarr commented on src/som/interpreter/actors/Actor.java:
```diff
@@ -231,15 +239,20 @@ public void doRunWithObjectSafepoints(final ActorProcessingThread t) {
       while (getCurrentMessagesOrCompleteExecution()) {
         processCurrentMessages(t, dbg);
       }
+
+      if (VmSettings.UNIFORM_TRACING || VmSettings.KOMPOS_TRACING) {
+        t.swapTracingBufferIfRequestedUnsync();
+      }
+      t.currentlyExecutingActor = null;
```
```suggestion
```

I think we to make `doRunWithObjectSafepoints` `private` and the `finally clause of `doRun` already does this.
What do you think?

smarr commented on src/som/interpreter/actors/ReceivedRootNode.java:
```diff
@@ -53,6 +55,13 @@ protected ReceivedRootNode(final SomLanguage language,
+    if (VmSettings.UNIFORM_TRACING) {
+      messageTracer = new RecordOneEvent(TraceRecord.MESSAGE);
+    }
+    if (VmSettings.RECEIVER_SIDE_TRACING) {
```
I'll assume the flag implies UNIFORM_TRACING.

smarr commented on src/som/interpreter/actors/RegisterOnPromiseNode.java:
```diff
@@ -71,28 +125,55 @@ public void register(final SPromise promise, final PromiseMessage msg,
     public void register(final SPromise promise, final PromiseMessage msg,
         final Actor current) {
 
+      // Output.println("try register onerror " + promise.hashCode() + " in "
+      // + TracingActivityThread.currentThread().getActivity().getId());
+
```
```suggestion
```

I assume this isn't needed any longer.

smarr commented on src/som/interpreter/actors/SPromise.java:
```diff
@@ -335,26 +369,327 @@ public SPromise getChainedPromiseUnsync() {
+    /**
+     * Stores Message Objects in a priority Queue, to bring the messages in the original order.
+     * Use resolveReplay method before resolving promise.
+     *
```
```suggestion
```

smarr commented on src/tools/concurrency/TracingActors.java:
```diff
@@ -61,6 +60,29 @@ public final int getActorId() {
+      if (VmSettings.SENDER_SIDE_TRACING) {
+        msg.getTracingNode().record(this.version);
+        this.version++;
+        // TODO maybe try to get the recording itself done outside the synchronized method
```
Is this TODO still current?

daumayr replied:
I don't think this would change much, in a single threaded execution it doesn't matter. In a multithreaded execution it might reduce contention of an actor a bit, e.g. for philosophers, but i don't think the performance would change much.

smarr commented on tests/replay/test.sh:
```diff
@@ -74,3 +124,59 @@ do
+  $SOM_DIR/som -t8 -G -JXmx1500m -vmd -r core-lib/Benchmarks/AsyncHarness.ns MultiParadigmBench 1000 0 100
+  echo ""
+  echo "========================================================"
+  echo ""
```
And we might want to reduce the code duplication by using a function.

smarr commented on core-lib/Benchmarks/AsyncHarness.ns:
```diff
@@ -120,9 +120,21 @@ class AsyncHarness usingPlatform: platform = Value (
+      | stats = system traceStatistics.
+        memstats = system memoryStatistics.
+      |
+
```
Can we avoid printing anything for which the value is 0?
I am not sure all our tools work with this, but it would minimize the output.

daumayr replied:
I implemented it for the trace stats, MemoryStats have to be enabled with a flag so that shouldn't be necessary

smarr reviewed (commented):
Ok, pfew, that took a while.

daumayr committed:
Update .gitlab-ci.yml

Co-authored-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
remove Comment

Co-authored-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
cleanup

daumayr committed:
remove Comment

Co-authored-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
Use ClassProfile for ContextNode

daumayr committed:
simplify test script

daumayr committed:
fix bug

One_Event_Size is not enough for the trace structuring events

daumayr committed:
print tracestats only when > 0

daumayr committed:
fix mutex command

daumayr committed:
update dynamic metrics

smarr commented:
With #369, we got the benchmarks already merged.
The current state of this code is pushed here, in case things go wrong: https://github.com/MetaConc/SOMns/tree/da/Uniform_Tracing_2020-08-08_Start

I'll update and force-push the branch for this PR now rebased on top of the merged #369.

smarr committed:
Fix use of old ACTOR_TRACING flag

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove redundant benchmarking

bench.conf is used for the paper, though, not in the repo, which means it will break and be not maintained. So, better to remove it.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove redundant code

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented on src/som/vmobjects/SSymbol.java:
```diff
@@ -41,10 +40,9 @@
-        || VmSettings.TRACK_SNAPSHOT_ENTITIES) {
+    if (VmSettings.KOMPOS_TRACING || VmSettings.TRACK_SNAPSHOT_ENTITIES) {
       symbolId = (short) idGenerator.getAndIncrement();
-      TracingBackend.logSymbol(this);
```
Hmm. Does this work for Kompos?
Does the logSymbol still need to be done for it?

smarr replied:
I restored the call. Seemed still necessary.

smarr committed:
Restore TracingBackend.logSymbol call for Kompos

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@daumayr so, to get the validation tests enabled again, we need to also handle the time threads? Do I remember that correctly?

I think, I prefer to actually merge this as is. It's huge already. And then try to do that in a new branch.
What do you think?

If that's ok for you, and my last changes are also fine with you, I'll merge this.

daumayr commented:
@smarr i had a look at your changes and they look ok. 
yes, to enable the validation tests we need to have a timer that is compatible with our event infrastructure, i.e. compatible with TracingActivityThread
